Remove changes.h, replace by bits

gassmoeller opened the pull request:
Fixes #1441. I looked at deal.II's workflow and it seemed a bit too complicated for our purposes, so I simplified it. So far we did not distinguish between major and minor changes, and the approach using cmake did not fit well our current shell script of `increment_version.sh`, therefore I decided to go with a shell script instead.
`increment_version.sh` works as before, and `release-tasklist` can stay untouched. Did I forget anything else?

bangerth commented on doc/modules/current_changes_header.template:
```diff
@@ -6,6 +6,3 @@
  * <ol>
- *
- * </ol>
- */
```
do we even still need this file, given that you've added `current_changes_header`?

tjhei commented on doc/modules/README.md:
```diff
@@ -0,0 +1,20 @@
+    <br>
+    (John Doe, YYYY/MM/DD)
+
+and is named "YYYYMMDD\_JohnDoe". File names for multiple contributions from the 
```
is there a reason for escaping the underscore here?

tjhei commented on doc/modules/create_changes.sh:
```diff
@@ -0,0 +1,38 @@
+
+if [ -e $output ]
+then
+  echo "ERROR: File '$output' already exists."
```
Why are you not simply overwriting the changes.h?

gassmoeller commented:
I think I addressed all comments, but I am not very  familiar with the Makefile syntax, so please have a look.
About your questions:
1. We still need current_changes_header.template, because that is the generic header, which is specialized for the latest release by `increment_version.sh` into `current_changes_header`. Otherwise `create_changes.h` can not figure out the last release (VERSION is already at the next one), or we would need to modify `current_changes_header` by hand after every release.
2. I do not think the underscore is _necessary_ in this case, but my vi complained about it, because two underscores in markdown will render the text in between italic. Do you think it confuses readers?

tjhei commented:
> Do you think it confuses readers?

Maybe. Can you write the filename as ``YYYYMMDD_JohnDoe`` (in double backticks but without escaping the underscore).

gassmoeller commented:
Yes double backticks seem to work and are less confusing. I changed the file accordingly.

gassmoeller commented:
Ok the tester has passed and I have Timos ok, so I will merge. I mostly want to avoid other PRs getting merged before this one, because this PR needs to be updated every time another one adds something to changes.h. Of course this means every open PR with entries in changes.h will need to be rebased, but they would probably have to do that anyway.

bangerth commented on doc/Makefile:
```diff
@@ -30,11 +31,15 @@ all: aspect.tag manual.pdf
 	@for i in `seq 1 10`; do sed -i 's/{\([^!]*\)!\([^!]*\)!\([^!]*\)!\([^}]*\)}/{\1!\2!\3\/\4}/' $@; done
 
+# Generate changes.h from the bits in modules/changes/
+modules/changes.h: modules/changes/* modules/create_changes.sh modules/current_changes_*
```
Looking through this again, this won't work -- `make` does not expand wildcards. You'll have to write `$(shell echo modules/changes/* modules/create_changes.sh modules/current_changes_*)`

gassmoeller replied:
Are you sure about this? https://www.gnu.org/software/make/manual/html_node/Wildcards.html suggests wildcards in targets and prerequisites are expanded automatically, and I tested touching a random file in `modules/changes/` and it recreated changes.h. Maybe `make` did learn a thing or two over its (long) lifetime :smile:.

bangerth replied:
Ah, fantastic -- it works in targest and prerequisites. I had only remembered the case where it is used elsewhere, such as in the last example at https://www.gnu.org/software/make/manual/html_node/Wildcard-Examples.html#Wildcard-Examples

In general, wildcards for prerequisites is dangerous because it only picks up files *that are already in the directory*. It won't pick up files that are created by another rule, and because it doesn't pick them up, they won't be generated. In your case, none of the listed prerequisites fall in this category, so this is not an issue.
